### DISCUSSION TITLE
RTC clock integration

### DISCUSSION THREAD

**Discussion by min0me** [bxparks/AceTime#67] [Q&A]
Hello,
I am trying to add the AceTime library into a project to make a timer DST aware and struggling somewhat.

I don't think the library is detecting the RTC correctly . Using another library with the standard i2c pinout I am able to detect the RTC and serial print the time/date from it and it is correct.

I am using a DS3231 (DS3231SN)  RTC for now (there is an atmel eeprom on there too) and have tried with a DOIT ESP32 DEVKIT v1 / Wemos LOLIN32 lite/ Wemos D1 (ESP8266) clone/ Arduino UNO along with V1.7.2

I am using the example found in the documentation (it can be a little bit tricky to follow with no comments within the code as the namespaces are a new concept to me.).

``` c++
#include <AceTime.h>
using namespace ace_time;
using namespace ace_time::clock;

DS3231Clock dsClock;

void setup() {
  Serial.begin(115200);
  while(!Serial); // needed for Leonardo/Micro

  dsClock.setup();
//  dsClock.setNow(0); // 2000-01-01T00:00:00Z comment out as i don't want to change the rtc time
}

void loop() {
  Serial.print (dsClock.getNow()); // print epoch value to see if rtc is working correctly
  acetime_t nowSeconds = dsClock.getNow();
  // convert epochSeconds to UTC-08:00
  OffsetDateTime odt = OffsetDateTime::forEpochSeconds(
      nowSeconds, TimeOffset::forHours(8));
  odt.printTo(Serial);
  delay(10000); // wait 10 seconds
}
```

I get a static value for the epoch time that does not look correct -2147483648
Additionally i get an ` <Invalid OffsetDateTime> `

I am unsure where I am going wrong, but any advise would be awesome

Regards,

**Comment by bxparks** [comment]
Try adding the following at the top of your code:

```C++
#include <Wire.h>
```

Then modify your `setup()` to add a `Wire.begin()`:

```C++
void setup() {
  Serial.begin(115200);
  while(!Serial); // needed for Leonardo/Micro

  Wire.begin(); // <---------------------------- add this

  dsClock.setup();
}
```

The `DS3231Clock` uses the pre-installed `<Wire.h>` library to communicate with the DS3231 chip over I2C. So you have to initialize the Wire library explicitly. The reason that `dsClock.setup()` does not do this for you is because there are different ways to initialize the I2C, and it would not be appropriate for the `DS3231Clock` class to assume that the user wants to initialize the `<Wire.h>` library in a certain way.

Looking at my `examples/` directory, it is clear to me that I should add more example code in there. It used to have more, but many of them grew in complexity, and became actual independent applications, so got moved into the https://github.com/bxparks/clocks project. Those clock projects are now so complex and feature rich, they are no longer serve as good samples to use as example code. I hope to fix this situation in the near future in an upcoming major refactoring of the code. Hope this helps.

**Reply by min0me** [reply]
Hi,
Thanks that has solved it. I thought I read somewhere that the wire lib was already called by calling acetime, but I could be that I saw it done in a fork and got confused.

Yes, more examples couldn't hurt, but some very basic ones with heavy commenting throughout would be awesome. Its hard to follow if they have everything including the kitchen sink in them.

I will have a dig in the clocks repo and see if i can find more examples to give me an idea what i am trying to build.

Thank again
